api/backend/multibackend.go

```diff
@@ -69,32 +63,48 @@ func NewParallelMultiBackend(backend api.Backend, limit int) api.MultiBackend {
 // fetchLazy issues a goroutine to compute the timeseries once a fetchticket becomes available.
 // It returns a channel to wait for the response to finish (the error).
 // It stores the result of the function invokation in the series pointer it is given.
-func (m *parallelMultiBackend) fetchLazy(result *api.Timeseries, fun func() (api.Timeseries, error), channel chan error) {
+func (m *parallelMultiBackend) fetchLazy(cancellable api.Cancellable, result *api.Timeseries, work func() (api.Timeseries, error), channel chan error) {
 	go func() {
-		ticket := <-m.tickets
-		series, err := fun()
-		// Put the ticket back (regardless of whether caller drops)
-		m.tickets <- ticket
-		// Store the result
-		*result = series
-		// Return the error (and sync up with the caller).
-		channel <- err
+		select {
+		case ticket := <-m.tickets:
+			series, err := work()
+			// Put the ticket back (regardless of whether caller drops)
+			m.tickets <- ticket
+			// Store the result
+			*result = series
+			// Return the error (and sync up with the caller).
+			channel <- err
+		case <-cancellable.Done():
+			channel <- api.BackendError{
+				api.TaggedMetric{},
+				api.FetchTimeoutError,
+				"",
+			}
+		}
 	}()
 }
 
 // fetchManyLazy abstracts upon fetchLazy so that looping over the resulting channels is not needed.
 // It returns any overall error, as well as a slice of the resulting timeseries.
-func (m *parallelMultiBackend) fetchManyLazy(funs []func() (api.Timeseries, error)) ([]api.Timeseries, error) {
-	results := make([]api.Timeseries, len(funs))
-	channel := make(chan error, len(funs)) // Buffering the channel means the goroutines won't need to wait.
+func (m *parallelMultiBackend) fetchManyLazy(cancellable api.Cancellable, works []func() (api.Timeseries, error)) ([]api.Timeseries, error) {
+	results := make([]api.Timeseries, len(works))
+	channel := make(chan error, len(works)) // Buffering the channel means the goroutines won't need to wait.
 	for i := range results {
-		m.fetchLazy(&results[i], funs[i], channel)
+		m.fetchLazy(cancellable, &results[i], works[i], channel)
 	}
 	var err error = nil
-	for _ = range funs {
-		thisErr := <-channel
-		if thisErr != nil {
-			err = thisErr
+	for _ = range works {
+		select {
+		case thisErr := <-channel:
+			if thisErr != nil {
+				err = thisErr
+			}
+		case <-cancellable.Done():
+			err = api.BackendError{
```

Comment:
It makes more sense to just return immediately, since it will match (nondeterministically) every remaining iteration of the loop.